src/ruby_supportlib/phusion_passenger/vendor/daemon_controller.rb

```diff
@@ -578,45 +551,88 @@ def log_file_has_changed?
-        diff = f.read.strip
-        if diff.empty?
-          nil
-        else
-          diff
-        end
+        f.read.strip
       end
-    else
-      nil
     end
   rescue Errno::ENOENT, Errno::ESPIPE
     # ESPIPE means the log file is a pipe.
     nil
   end
 
-  def determine_lock_file(options, identifier, pid_file)
-    if options[:lock_file]
-      LockFile.new(File.expand_path(options[:lock_file]))
+  def determine_lock_file(given_lock_file, identifier, pid_file)
+    if given_lock_file
+      LockFile.new(File.absolute_path(given_lock_file))
     else
-      LockFile.new(File.expand_path(pid_file + ".lock"))
+      LockFile.new(File.absolute_path(pid_file + ".lock"))
     end
   end
 
-  def self.fork_supported?
-    RUBY_PLATFORM != "java" && RUBY_PLATFORM !~ /win32/
-  end
-
-  def self.spawn_supported?
-    # Process.spawn doesn't work very well in JRuby.
-    Process.respond_to?(:spawn) && RUBY_PLATFORM != "java"
-  end
-
   def run_command(command)
     if should_capture_output_while_running_command?
-      run_command_while_capturing_output(command)
+      # Create tempfile for storing the command's output.
+      tempfile = Tempfile.new("daemon-output")
+      tempfile.chmod(0o666)
+      tempfile_path = tempfile.path
+      tempfile.close
+
+      spawn_options = {
+        in: "/dev/null",
+        out: tempfile_path,
+        err: tempfile_path,
+        close_others: true
+      }
     else
-      run_command_without_capturing_output(command)
+      spawn_options = {
+        in: "/dev/null",
+        out: :out,
+        err: :err,
+        close_others: true
+      }
+    end
+
+    if @keep_ios
+      @keep_ios.each do |io|
+        spawn_options[io] = io
+      end
+    end
+
+    pid = if @daemonize_for_me
+      Process.spawn(@env || {}, ruby_interpreter, SPAWNER_FILE,
+        command, spawn_options)
+    else
+      Process.spawn(@env || {}, command, spawn_options)
+    end
+
+    # run_command might be running in a timeout block (like
+    # in #start_without_locking).
+    begin
+      Process.waitpid(pid)
+    rescue Errno::ECHILD
+      # Maybe a background thread or whatever waitpid()'ed
+      # this child process before we had the chance. There's
+      # no way to obtain the exit status now. Assume that
+      # it started successfully; if it didn't we'll know
+      # that later by checking the PID file and by pinging
+      # it.
+      return InternalCommandOkResult.new(pid, tempfile_path ? File.read(tempfile_path).strip : nil)
+    rescue Timeout::Error
+      return InternalCommandTimeoutResult.new(pid, tempfile_path ? File.read(tempfile_path).strip : nil)
+    end
+
+    child_status = $?
+    output = File.read(tempfile_path).strip if tempfile_path
+    if child_status.success?
+      InternalCommandOkResult.new(pid, output)
+    else
+      InternalCommandErrorResult.new(pid, output, child_status)
+    end
+  ensure
+    begin
+      File.unlink(tempfile_path) if tempfile_path
```

Comment:
if you use `Tempfile.create` above, you can use `tempfile.unlink` here

Reply:
Thanks, I've upstreamed this: https://github.com/FooBarWidget/daemon_controller/commit/5b4223291619d23a9b7a14304c882e85e2025858